Composer/packages/client/src/components/ProjectTree/ProjectTree.tsx

```diff
@@ -55,28 +46,57 @@ const root = css`
   }
 `;
 
+const icons = {
+  TRIGGER: 'LightningBolt',
+  DIALOG: 'Org',
+  BOT: 'CubeShape',
+  EXTERNAL_SKILL: 'Globe',
+  FORM_DIALOG: '',
+  FORM_FIELD: 'Variable2', // x in parentheses
+  FORM_TRIGGER: 'TriggerAuto', // lightning bolt with gear
+  FILTER: 'Filter',
+};
+
+const tree = css`
+  height: 100%;
+  overflow-x: hidden;
+  overflow-y: auto;
+  height: 100%;
+  label: tree;
+`;
+
 // -------------------- ProjectTree -------------------- //
 
-function createGroupItem(dialog: DialogInfo, currentId: string, position: number, warningContent: string): IGroup {
-  return {
-    key: dialog.id,
-    name: dialog.displayName,
-    level: 1,
-    startIndex: position,
-    count: dialog.triggers.length,
-    hasMoreData: true,
-    isCollapsed: dialog.id !== currentId,
-    data: { ...dialog, warningContent },
-  };
+export type TreeLink = {
+  displayName: string;
+  isRoot: boolean;
+  warningContent?: string;
+  errorContent?: string;
+  projectId: string;
+  skillId: string | null;
+  dialogName?: string;
+  trigger?: number;
+};
+
+function isLinkEqual(link1: TreeLink | undefined, link2: TreeLink | undefined) {
+  if (link1 === link2) return true;
+  if (link1 == null) return false;
+  if (link2 == null) return false;
+  if (link1.projectId !== link2.projectId) return false;
+  if (link1.skillId !== link2.skillId) return false;
+  if (link1.dialogName !== link2.dialogName) return false;
+  if (link1.trigger !== link2.trigger) return false;
+  return true;
 }
 
-function createItem(trigger: ITrigger, index: number, warningContent: string) {
-  return {
-    ...trigger,
-    index,
-    warningContent,
-    displayName: trigger.displayName || getFriendlyName({ $kind: trigger.type }),
-  };
+export type TreeMenuItem = {
+  icon?: string;
+  label: string; // leave this blank to place a separator
+  action?: (link: TreeLink) => void;
```

Comment:
is this a click action? If yes, please name it accordingly. If not, shouldn't it be an array of actions?
```
type Action {
   label: string;
   onClick: (link: TreeLink) => void;
}
```

Reply:
No, you're right, it should be `onClick`.